src/SSymmetricCompact.jl

```diff
@@ -0,0 +1,269 @@
+    I = _symmetric_compact_indices(Val(N))
+    j, lower = I[i]
+    x′ = lower ? x : transpose(x)
+    return SSymmetricCompact{N}(setindex(a.lowertriangle, x′, j))
+end
+
+# needed because it is used in convert.jl and the generic fallback is slow
+@generated function Base.Tuple(a::SSymmetricCompact{N}) where N
+    exprs = [:(a[$i]) for i = 1 : N^2]
+    quote
+        @_inline_meta
+        tuple($(exprs...))
+    end
+end
+
+LinearAlgebra.ishermitian(a::SSymmetricCompact{N, T}) where {N,T <: Real} = true
+LinearAlgebra.ishermitian(a::SSymmetricCompact) = all(isreal, a.lowertriangle)
+LinearAlgebra.issymmetric(a::SSymmetricCompact) = true
+
+# TODO: factorize?
+
+# TODO: a.lowertriangle == b.lowertriangle is slow (used by SDiagonal). SMatrix etc. actually use AbstractArray fallback (also slow)
+@inline Base.:(==)(a::SSymmetricCompact, b::SSymmetricCompact) = mapreduce(==, (x, y) -> x && y, a.lowertriangle, b.lowertriangle)
+@generated function _map(f, a::SSymmetricCompact...)
+    S = Size(a[1])
+    N = S[1]
+    L = triangularnumber(N)
+    exprs = Vector{Expr}(undef, L)
+    for i ∈ 1:L
+        tmp = [:(a[$j].lowertriangle[$i]) for j ∈ 1:length(a)]
+        exprs[i] = :(f($(tmp...)))
+    end
+    return quote
+        @_inline_meta
+        same_size(a...)
+        @inbounds return SSymmetricCompact(SVector(tuple($(exprs...))))
+    end
+end
+
+# Scalar-array. TODO: overload broadcast instead, once API has stabilized a bit
+@inline Base.:+(a::Number, b::SSymmetricCompact) = SSymmetricCompact(a + b.lowertriangle)
+@inline Base.:+(a::SSymmetricCompact, b::Number) = SSymmetricCompact(a.lowertriangle + b)
+
+@inline Base.:-(a::Number, b::SSymmetricCompact) = SSymmetricCompact(a - b.lowertriangle)
+@inline Base.:-(a::SSymmetricCompact, b::Number) = SSymmetricCompact(a.lowertriangle - b)
+
+@inline Base.:*(a::Number, b::SSymmetricCompact) = SSymmetricCompact(a * b.lowertriangle)
+@inline Base.:*(a::SSymmetricCompact, b::Number) = SSymmetricCompact(a.lowertriangle * b)
+
+@inline Base.:/(a::SSymmetricCompact, b::Number) = SSymmetricCompact(a.lowertriangle / b)
+@inline Base.:\(a::Number, b::SSymmetricCompact) = SSymmetricCompact(a \ b.lowertriangle)
+
+@generated function _plus_uniform(::Size{S}, a::SSymmetricCompact{N, T, L}, λ) where {S, N, T, L}
+    @assert S[1] == N
+    @assert S[2] == N
+    exprs = Vector{Expr}(undef, L)
+    i = 0
+    for col = 1 : N, row = col : N
+        i += 1
+        exprs[i] = row == col ? :(a.lowertriangle[$i] + λ) : :(a.lowertriangle[$i])
+    end
+    return quote
+        @_inline_meta
+        R = promote_type(eltype(a), typeof(λ))
+        SSymmetricCompact{N, R, L}(SVector{L, R}(tuple($(exprs...))))
+    end
+end
+
+@generated function LinearAlgebra.transpose(a::SSymmetricCompact{N, T, L}) where {N, T, L}
+    # For M = [A Bᵀ
+    #          B C],
+    # Mᵀ = [Aᵀ Bᵀ
+    #       B  Cᵀ]
+    # In general, we should only recursively transpose the diagonal elements.
+    exprs = Vector{Expr}(undef, L)
```

Comment:
I don't get this comment

Reply:
One might think that you should transpose all of the elements of `a.lowertriangle` (which might e.g. be of type `Matrix{ComplexF64}`) to match the recursive property of `transpose` in the `LinearAlgebra` implementation, as I did initially. But that's not the case. This comment is trying to explain that. Suggestions to make it clearer?

Reply:
I had to think about this myself when I looked through the code yesterday.
It seems mysterious, but it's only because a symmetric block matrix may not be a symmetric matrix when it's flattened.

```
# Transposition of a symmetric matrix is the identity.
# However, if we happen to be storing a block matrix there may be
# non-symmetric diagonal parts so the blocks on the diagonal need
# transposing:
# [A  Bᵀ] ᵀ  =  [Aᵀ  Bᵀ]  =  [Aᵀ Bᵀ]
# [B  C ]       [Bᵀᵀ Cᵀ]     [B  Cᵀ]
```

Reply:
Done.